.changeset/mighty-phones-press.md

```diff
@@ -0,0 +1,5 @@
+---
+'xstate': patch
+---
+
+During external events, only enter target plus descendent nodes.
```

Comment:
Unfortunately, this is not correct. The external transition should always leave the source state. If the target is an ancestor of the source then everything up to that ancestor (including that ancestor) should be exited, target should be reentered and initial states etc should be recursively resolved and entered.

Reply:
It is exiting the source node correctly. The issue is that it's entering all the ancestor nodes until it reaches the target. In the example provided in the issue here https://github.com/statelyai/xstate/issues/3618, it was only mentioned that it re-entered the source node, but if you were to add entry actions in the other nodes, you would see that all of them are being triggered until it reached the source.

Reply:
Ah, yes - that sounds like the issue that i’ve identified in that linked issue. However, the fix seems to be incomplete (I might be wrong). Perhaps it works because other code “patches” the result of this.

Once you add a test case I will take a second look at this and gonna try to figure out a test case that would prove this being incomplete, or I will land this.

thank you for your contribution! It’s very much appreciated